dockerGrader/testServer.py

```diff
@@ -0,0 +1,121 @@
+        f.write(json.dumps(submission, indent=2))
+
+def uploadResult(project, studentID):
+    if not os.path.exists("{}/result.json".format(TEST_DIR)):
+        return False
+    with open("result.json", "r") as fr:
+        serializedResult = fr.read()
+    s3.put_object(
+        Bucket=BUCKET,
+        Key='ta/grading/{}/{}.txt'.format(project, studentID),
+        Body=serializedResult.encode('utf-8'),
+        ContentType='text/plain')
+    return True
+
+def lookupGoogleId(netId):
+    path = 'users/net_id_to_google/%s.txt' % netId
+    try:
+        response = s3.get_object(Bucket=BUCKET, Key=path)
+        net_id = response['Body'].read().decode('utf-8')
+        return net_id
+    except botocore.exceptions.ClientError as e:
+        if not e.response['Error']['Code'] == "NoSuchKey":
+            # unexpected error
+            logging.warning(
+                'Unexpected error when look up Googlg ID:' + e.response['Error']['Code'])
+        raise e
+
+def fetchFromS3(project, netId):
+    googleId = lookupGoogleId(netId)
+    if not googleId:
+        return None
+    curPath = ACCESS_PATH.format(project=project, googleId=googleId)
+    downloadSubmission(curPath)
+
+def sendToDocker(project, netId):
+    # create directory to mount inside a docker container
+    if os.path.exists(TEST_DIR):
+        shutil.rmtree(TEST_DIR)
+    shutil.copytree(CODE_DIR, TEST_DIR)
+    shutil.copytree("./io", TEST_DIR + "/io")
+    # we can't use shutil.copytree here again because TEST_DIR exists
+    testCodePath = "./test/{}".format(project)
+    for item in os.listdir(testCodePath):
+        src = os.path.join(testCodePath, item)
+        dest = os.path.join(TEST_DIR, item)
+        if os.path.isdir(src):
+            shutil.copytree(src, dest)
+        else:
+            shutil.copy2(src, dest)
+
+    # run tests inside a docker container
+    image = 'python:3.7-stretch' # TODO: find/build some anaconda image
+
+    cmd = ['docker', 'run',                           # start a container
+           '-v', os.path.abspath(TEST_DIR)+':/code',  # share the test dir inside
+           '-w', '/code',                             # working dir is w/ code
+           image,                                     # what docker image?
+           'python3', 'testMain.py',
+           '-p', project,
+           '-i', netId]                      # command to run inside
+    logging.info("docker cmd:" + ' '.join(cmd))
+    output = subprocess.check_output(cmd)
+    with open(TEST_DIR + '/result.json') as fr:
+        result = fr.read()
+    return result
+
+@app.route('/')
+def index():
+    return "index"
+
+@app.route('/json/<project>/<netId>')
+def gradingJson(project, netId):
+    fetchFromS3(project, netId)
+    sendToDocker(project, netId)
+    return ""
+    try:
+        fetchFromS3(project, netId)
+        sendToDocker(project, netId)
+        return json.dumps({})
+    except Exception as e:
+        return json.dumps({"error": str(e)})
```

Comment:
Maybe this should only go to the log?  Currently, people could use this to determine who has turned something in.